### PR TITLE
cpu/msp430_common: add flashpage_raw

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

 This PR adds `flashpage_raw` for msp430. I specified the type for `FLASHPAGE_SIZE` and `FLASHPAGE_NUMOF` since `FLASHPAGE_SIZE*FLASHPAGE_NUMOF` would overflow if configured as int on 16bit's msp430.

This PR also modifies `tests/mtd_flashpage` and `tests/periph_flashpage` to adapt to msp430 and 16bit.

There where some issue in the available flash specified in `cpu_conf` and in TI official doc, see abfc189 for more details.

### Testing procedure

Tested on `z1`:

`BUILD_IN_DOCKER=1 make -C tests/periph_flashpage/ BOARD=z1 flash test`

`BUILD_IN_DOCKER=1 make -C tests/mtd_flashpage/ BOARD=z1 flash test`

_note: I had to reset my zollertia manually since the debugger didn't seem to be doing it._

### CODE REVIEW

File: tests/periph_flashpage/main.c [hunk 1]
```diff
@@ -196,22 +196,33 @@ static uint32_t getaddr(const char *str)
 
 static int cmd_write_raw(int argc, char **argv)
 {
+#ifdef CPU_MSP430FXYZ
```
Comment (benpicco) [inline tests/periph_flashpage/main.c]: Can we use

`#if __SIZEOF_POINTER__ == 2`

instead for a more general solution?

Comment (fjmolinas) [reply tests/periph_flashpage/main.c]: Thanks @benpicco!, I wanted something like! Addressed

**Comment by fjmolinas** [comment]
@benpicco are current changes good for you? Do you have a board to test?

**Comment by benpicco** [comment]
@fjmolinas The changes look good to me, but I don't have a board to test.

**Comment by benpicco** [comment]
Feel free to squash - so I guess we now need a second maintainer with the hardware to give the ACK?

**Comment by fjmolinas** [comment]
@aabadie could you test this? I can lend you the zolertia I have.

**Comment by aabadie** [comment]
>  I can lend you the zolertia I have.

That will help, thanks.

**Comment by fjmolinas** [comment]
@aabadie @benpicco rebased and squashed!

**Comment by fjmolinas** [comment]
Seems I screwed the rebase...

**Comment by fjmolinas** [comment]
Somehow squashing over-did a change... I re-triggered to see the mtd-flash issues.

**Comment by fjmolinas** [comment]
To options with `mtd` issues:

* Blacklist 16bit msp430
* Change `mtd` api

**Comment by benpicco** [comment]
Or how about 

```C
void __attribute__((weak)) flashpage_read_raw(const void *buf, uint32_t addr, uint32_t size)
{
    memcpy(buf, (void*)addr, size);
}
```

and then have a platform specific implementation for msp430.

**Comment by fjmolinas** [comment]
> ```
> void __attribute__((weak)) flashpage_read_raw(const void *buf, uint32_t addr, uint32_t size)
> {
>     memcpy(buf, (void*)addr, size);
> }
> ```

So extend the `flashpage_raw`? This is actually something I wanted to do and remove  the dependency of `flashpage_raw` with `flashpage`.

> But yes, it's probably better if those flashpage functions pass around addresses instead of pointers.

You would change `flashpage` api then? I was thinking of changing the api for `mtd` to receive `(void*)` as address so it cant get casted according to size. Whats the argument for one over the other? 

If we would got with `flashpage functions pass around addresses instead of pointers` then I would ave this as a second step and not in this PR.

**Comment by benpicco** [comment]
I just briefly looked into `mtd_flashpage.c`.
If I understood correctly, this is to use the internal flash through the same API as you would for e.g. SPI flash.

So you would pass addresses anyway, I'd say just pass them on to the actual platform specific implementation, you can still cast them to a pointer there - and have the possibility to do something else if the platform requires it.

File: drivers/mtd_flashpage/mtd_flashpage.c [hunk 2]
```diff
@@ -27,7 +27,9 @@
 #include "mtd_flashpage.h"
 #include "periph/flashpage.h"
 
-#define MTD_FLASHPAGE_END_ADDR     CPU_FLASH_BASE + (FLASHPAGE_NUMOF * FLASHPAGE_SIZE)
+/* Avoid 16bit overflow, use start addres of the last byte */
```
Comment (aabadie) [inline drivers/mtd_flashpage/mtd_flashpage.c]: s/address/addres**s**/

**Comment by fjmolinas** [comment]
@aabadie I fixed the change that broke for the other board. Still doesn't work on zolertia, partly because its flash it's not aligned to page sizes.. I'll need to debug some more.

File: drivers/mtd_flashpage/mtd_flashpage.c [hunk 3]
```diff
@@ -27,7 +27,7 @@
 #include "mtd_flashpage.h"
 #include "periph/flashpage.h"
 
-#define MTD_FLASHPAGE_END_ADDR     CPU_FLASH_BASE + (FLASHPAGE_NUMOF * FLASHPAGE_SIZE)
+#define MTD_FLASHPAGE_END_ADDR     (CPU_FLASH_BASE + ((FLASHPAGE_NUMOF) * FLASHPAGE_SIZE))
```
Comment (aabadie) [inline drivers/mtd_flashpage/mtd_flashpage.c]: unrelated change ?

**Comment by fjmolinas** [comment]
I've discovered multiple issue, so tagging as wip.

**Comment by fjmolinas** [comment]
Here are the issues encountered:

- TI datasheet are actually faulty and `msp430f1612` and `msp430f2617` have actually `54.75kb` and `91.75kb`.
- Our code for `msp430f1612` is discarding the first "sector"('page') without any documentation and specifies an improper number of pages for this to be accurate
- Our code for `msp43021612` mentions that we only have access to `52kb` (because we support 16bit and not 20bit) but doesn't define the number of pages accordingly.

I can easily fix these issues, and I would think what was done (although not documented) for `msp430f1612` is the right way to go since we don't have support for varied sized page, and it only means loosing the first sector of 256bytes. So I would like to:

- define `FLASHPAGE_NUMOFF` as only the availabe pages
- define `CPU_FLASHBASE` to skip the first sector of 256bytes
- proper comments for this
- limit linking addresses (??)

My issue is that we are limiting the range for `flashpage` while linking is not, I think that to fix this properly the address range used for linking should be limited. This I'm not sure how to do.

**Comment by fjmolinas** [comment]
> - define FLASHPAGE_NUMOFF as only the availabe pages
> - define CPU_FLASHBASE to skip the first sector of 256bytes
> - proper comments for this
> - limit linking addresses (??)

@aabadie I went ahead with points 1 to 3, I would like to postpone linking until after `msp430` toolchain change is done. Would that be ok with you?

**Comment by fjmolinas** [comment]
@aabadie ping!

File: tests/mtd_flashpage/main.c [hunk 4]
```diff
@@ -19,8 +19,22 @@
 #include "mtd.h"
 #include "mtd_flashpage.h"
 
-#define TEST_ADDRESS1       (uint32_t)flashpage_addr(FLASHPAGE_NUMOF - 1)
-#define TEST_ADDRESS2       (uint32_t)flashpage_addr(FLASHPAGE_NUMOF - 2)
+/* For MSP430 cpu's the last page holds the interrupt vector, allthough the api
```
Comment (aabadie) [inline tests/mtd_flashpage/main.c]:
```suggestion
/* For MSP430 cpu's the last page holds the interrupt vector, although the api
```

File: tests/mtd_flashpage/main.c [hunk 5]
```diff
@@ -19,8 +19,22 @@
...
+   should not limit erasing that page, we don't want to brake when testing */
```
Comment (aabadie) [inline tests/mtd_flashpage/main.c]:
```suggestion
   should not limit erasing that page, we don't want to break when testing */
```

File: tests/periph_flashpage/main.c [hunk 6]
```diff
@@ -28,6 +28,14 @@
 
 #define LINE_LEN            (16)
 
+/* For MSP430 cpu's the last page holds the interrupt vector, allthough the api
```
Comment (aabadie) [inline tests/periph_flashpage/main.c]:
```suggestion
/* For MSP430 cpu's the last page holds the interrupt vector, although the api
```

File: tests/periph_flashpage/main.c [hunk 7]
```diff
@@ -28,6 +28,14 @@
...
+   should not limit erasing that page, we don't want to brake when testing */
```
Comment (aabadie) [inline tests/periph_flashpage/main.c]:
```suggestion
   should not limit erasing that page, we don't want to break when testing */
```

File: drivers/mtd_flashpage/mtd_flashpage.c [hunk 8]
```diff
@@ -38,21 +38,29 @@ static int _init(mtd_dev_t *dev)
 
 static int _read(mtd_dev_t *dev, void *buf, uint32_t addr, uint32_t size)
 {
-    assert(addr < MTD_FLASHPAGE_END_ADDR);
+    assert(addr <= MTD_FLASHPAGE_END_ADDR);
```
Comment (aabadie) [inline drivers/mtd_flashpage/mtd_flashpage.c]: Are you sure this is needed ?

**Comment by aabadie** [comment]
Tested on z1 and nucleo-l073rz and it works. I have some minor comments but it looks good in general.

File: drivers/mtd_flashpage/mtd_flashpage.c (hunk 8 above)
Comment (fjmolinas) [reply drivers/mtd_flashpage/mtd_flashpage.c]: Got in by mistake, removed.

File: tests/periph_flashpage/main.c (hunk 7 above)
Comment (fjmolinas) [reply tests/periph_flashpage/main.c]: addressed.

File: tests/periph_flashpage/main.c (hunk 6 above)
Comment (fjmolinas) [reply tests/periph_flashpage/main.c]: addressed.

File: tests/mtd_flashpage/main.c (hunk 5 above)
Comment (fjmolinas) [reply tests/mtd_flashpage/main.c]: addressed.

File: tests/mtd_flashpage/main.c (hunk 4 above)
Comment (fjmolinas) [reply tests/mtd_flashpage/main.c]: addressed.

**Comment by fjmolinas** [comment]
@aabadie comments addressed, is it OK to squash?

**Comment by fjmolinas** [comment]
@aabadie @benpicco Thanks for the review.
